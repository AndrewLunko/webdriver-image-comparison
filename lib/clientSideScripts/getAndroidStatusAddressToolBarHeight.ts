import {StatusAddressToolBarHeight} from './statusAddressToolBarHeight.interfaces';
import {AndroidOffsets} from '../helpers/constants.interfaces';

/**
 * Get the current height of the Android status and address bar
 */
export default function getAndroidStatusAddressToolBarHeight(
  androidOffsets: AndroidOffsets
): StatusAddressToolBarHeight {
  // Determine version for the right offsets
  const {height, width} = window.screen;
  const {innerHeight} = window;
  const match = (navigator.appVersion).match(/Android (\d+).(\d+).?(\d+)?/);
  const majorVersion = parseInt(match[1], 10);
  const versionOffsets = androidOffsets[majorVersion];
  const statusAddressBarHeight = versionOffsets.STATUS_BAR + versionOffsets.ADDRESS_BAR;
  const toolBarHeight = height - innerHeight - statusAddressBarHeight;

  // Determine status, address and tool bar height
  return {
    statusAddressBar: {
      height: statusAddressBarHeight,
      width,
      x: 0,
      y: 0,
    },
    toolBar: {
      height: toolBarHeight,
      width,
      x: 0,
      y: height - toolBarHeight,
    },
  };
}
