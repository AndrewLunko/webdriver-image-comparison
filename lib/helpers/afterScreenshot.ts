import hideScrollBars from '../clientSideScripts/hideScrollbars';
import removeCustomCss from '../clientSideScripts/removeCustomCss';
import {CUSTOM_CSS_ID} from './constants';
import {formatFileName, getAndCreatePath} from './utils';
import {saveBase64Image} from '../methods/images';
import {join} from 'path';
import {Executor} from '../methods/methods.interface';
import {AfterScreenshotOptions, ScreenshotOutput} from './afterScreenshot.interfaces';
import hideRemoveElements from '../clientSideScripts/hideRemoveElements';

/**
 * Methods that need to be executed after a screenshot has been taken
 * to set all back to the original state
 */
export default async function afterScreenshot(executor: Executor, options: AfterScreenshotOptions): Promise<ScreenshotOutput> {
  const {
    actualFolder,
    base64Image,
    fileName: fileNameOptions,
    filePath,
    hideElements,
    hideScrollBars: noScrollBars,
    removeElements,
  } = options;

  // Get the path
  const path = getAndCreatePath(actualFolder, filePath);

  // Get the filePath
  const fileName = formatFileName(fileNameOptions);

  // Save the element
  await saveBase64Image(base64Image, join(path, fileName));

  // Show the scrollbars again
  await executor(hideScrollBars, !noScrollBars);

  // Show elements again
  if (hideElements.length > 0 || removeElements.length > 0) {
    await executor(hideRemoveElements, {hide: hideElements, remove: removeElements}, false);
  }

  // Remove the custom set css
  await executor(removeCustomCss, CUSTOM_CSS_ID);

  // Return the needed data
  return {
    fileName,
    path,
    devicePixelRatio: fileNameOptions.devicePixelRatio,
  };
}
