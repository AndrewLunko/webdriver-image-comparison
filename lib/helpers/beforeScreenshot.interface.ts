import {EnrichedInstanceData, InstanceData} from '../methods/instanceData.interfaces';

export interface BeforeScreenshotOptions {
  // The instance data
  instanceData: InstanceData;
  // The padding that needs to be added to the address bar on iOS and Android to do a proper cutout of the the viewport.
  addressBarShadowPadding: number;
  // Disable all css animations
  disableCSSAnimation: boolean;
  // Hide all scrollbars
  noScrollBars: boolean;
  // The padding that needs to be added to the tool bar on iOS and Android
  toolBarShadowPadding: number;
}

export interface BeforeScreenshotResult extends EnrichedInstanceData {};
