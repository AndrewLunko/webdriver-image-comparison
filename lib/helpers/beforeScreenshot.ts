import hideScrollBars from '../clientSideScripts/hideScrollbars';
import setCustomCss from '../clientSideScripts/setCustomCss';
import {CUSTOM_CSS_ID} from './constants';
import {getAddressBarShadowPadding, getToolBarShadowPadding} from './utils';
import getEnrichedInstanceData from '../methods/instanceData';
import {BeforeScreenshotOptions, BeforeScreenshotResult} from './beforeScreenshot.interface';
import {Executor} from '../methods/methods.interface';
import hideRemoveElements from '../clientSideScripts/hideRemoveElements';

/**
 * Methods that need to be executed before a screenshot will be taken
 */
export default async function beforeScreenshot(
  executor: Executor,
  options: BeforeScreenshotOptions,
  addShadowPadding: boolean = false
): Promise<BeforeScreenshotResult> {

  const {browserName, nativeWebScreenshot, platformName} = options.instanceData;
  const {
    addressBarShadowPadding,
    disableCSSAnimation,
    hideElements,
    noScrollBars,
    removeElements,
    toolBarShadowPadding,
  } = options;
  const addressBarPadding = getAddressBarShadowPadding({
    platformName,
    browserName,
    nativeWebScreenshot,
    addressBarShadowPadding,
    addShadowPadding,
  });
  const toolBarPadding = getToolBarShadowPadding({platformName, browserName, toolBarShadowPadding, addShadowPadding});

  // Hide the scrollbars
  await executor(hideScrollBars, noScrollBars);

  // Hide and or Remove elements
  if (hideElements.length > 0 || removeElements.length > 0) {
    await executor(hideRemoveElements, {hide: hideElements, remove: removeElements}, true);
  }

  // Set some custom css
  await executor(setCustomCss, {addressBarPadding, disableCSSAnimation, id: CUSTOM_CSS_ID, toolBarPadding});

  // Get all the needed instance data
  const instanceOptions = {
    addressBarShadowPadding: options.addressBarShadowPadding,
    toolBarShadowPadding: options.toolBarShadowPadding,
    ...(options.instanceData),
  };

  return getEnrichedInstanceData(executor, instanceOptions, addShadowPadding);
}
