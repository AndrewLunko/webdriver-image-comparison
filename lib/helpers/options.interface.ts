export interface ClassOptions{
  /**
   * Class options
   */
  addressBarShadowPadding?: number;
  autoSaveBaseline?: boolean;
  baselineFolder?: string;
  debug?: boolean;
  formatImageName?: string;
  nativeWebScreenshot?: boolean;
  savePerInstance?: boolean;
  screenshotPath?: string;
  toolBarShadowPadding?: number;

  /**
   * Class and method options
   */
  disableCSSAnimation?: boolean;
  fullPageScrollTimeout?: number;
  hideScrollBars?: boolean;

  /**
   * Compare options
   */
  blockOutStatusBar?: boolean;
  blockOutToolBar?: boolean;
  ignoreAlpha?: boolean;
  ignoreAntialiasing?: boolean;
  ignoreColors?: boolean;
  ignoreLess?: boolean;
  ignoreNothing?: boolean;
  ignoreTransparentPixel?: boolean;
  rawMisMatchPercentage?: boolean;
  returnAllCompareData?: boolean;
  saveAboveTolerance?: number;
}

export interface DefaultOptions {
  addressBarShadowPadding: number;
  autoSaveBaseline: boolean;
  debug: boolean;
  formatImageName: string;
  nativeWebScreenshot: boolean;
  savePerInstance: boolean;
  toolBarShadowPadding: number;
  resizeDimensions: {
    bottom: number;
    left: number;
    right: number;
    top: number;
  };
  disableCSSAnimation: boolean;
  fullPageScrollTimeout: number;
  hideScrollBars: boolean;
  compareOptions: CompareOptions
}

interface CompareOptions {
  blockOutStatusBar: boolean;
  blockOutToolBar: boolean;
  ignoreAlpha: boolean;
  ignoreAntialiasing: boolean;
  ignoreColors: boolean;
  ignoreLess: boolean;
  ignoreNothing: boolean;
  ignoreTransparentPixel: boolean;
  rawMisMatchPercentage: boolean;
  returnAllCompareData: boolean;
  saveAboveTolerance: number;
}
