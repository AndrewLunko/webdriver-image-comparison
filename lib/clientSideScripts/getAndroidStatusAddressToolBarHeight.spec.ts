import {OFFSETS} from '../helpers/constants';
import {ANDROID_DEVICES, NAVIGATOR_APP_VERSIONS} from '../mocks/mocks';
import getAndroidStatusAddressToolBarHeight from './getAndroidStatusAddressToolBarHeight';

describe('getAndroidStatusAddressToolBarHeight', () => {
  it('should the the android status, address and toolbar height', () => {
    Object.defineProperty(navigator, 'appVersion', {value: NAVIGATOR_APP_VERSIONS.ANDROID['9']});
    Object.defineProperty(window.screen, 'width', {value: ANDROID_DEVICES.NEXUS_5X.width});
    Object.defineProperty(window.screen, 'height', {value: ANDROID_DEVICES.NEXUS_5X.height});
    Object.defineProperty(window, 'innerHeight', {value: ANDROID_DEVICES.NEXUS_5X.innerHeight});

    expect(getAndroidStatusAddressToolBarHeight(OFFSETS.ANDROID)).toMatchSnapshot();
  });
});
