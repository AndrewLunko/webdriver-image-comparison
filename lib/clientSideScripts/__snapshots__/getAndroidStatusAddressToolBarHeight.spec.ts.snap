// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAndroidStatusAddressToolBarHeight should the the android status, address and toolbar height 1`] = `
Object {
  "statusAddressBar": Object {
    "height": 80,
    "width": 412,
    "x": 0,
    "y": 0,
  },
  "toolBar": Object {
    "height": 48,
    "width": 412,
    "x": 0,
    "y": 684,
  },
}
`;
